File: Java/soln-string-problems/anagrams-group-string-list.java
/**
 @author Farheen Bano
  
 References-
 https://www.geeksforgeeks.org/given-a-sequence-of-words-print-all-anagrams-together/
 https://leetcode.com/problems/group-anagrams/
*/

import java.util.*;
import java.lang.*;
import java.io.*;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        if (strs.length == 0) 
            return new ArrayList();
        Map<String, List> output = new HashMap<String, List>();
        
        for (String s : strs) {
            char[] ca = s.toCharArray();
            Arrays.sort(ca);
            String key = String.valueOf(ca);
            if (!output.containsKey(key)) 
                output.put(key, new ArrayList());
        
            output.get(key).add(s);
        }
        return new ArrayList(output.values());
    }
}

File: C++/soln-string-problems/anagrams-group-string-list.cpp
/**
 @author Farheen Bano

 Date 15-07-2021
  
 References-
 https://www.geeksforgeeks.org/given-a-sequence-of-words-print-all-anagrams-together/
 https://leetcode.com/problems/group-anagrams/
*/

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> result; 
        if (strs.size() == 0) 
            return result;
        
        unordered_map<string, vector<string>> anagrams;
        for (string s : strs) {
            string copy_s=s;
            sort(s.begin(),s.end());
            anagrams[s].push_back(copy_s);
        }
    
        for(auto x:anagrams)
            result.push_back(x.second);

        return result;
    }
};